### PR TITLE
Adding GEP 2907: TLS Configuration Placement and Terminology

### LINKED ISSUE #2907: GEP: TLS Configuration Placement and Terminology
**What would you like to be added**:
A memorandum GEP that provides a high level vision for TLS configuration in Gateway API. This GEP will not include any specific API changes, but will propose naming patterns for top level fields/buckets that should be used to TLS configuration.

**Why this is needed**:
We've got two TLS GEPs in flight (#91 and #1897), and likely more on the way. If we're not careful, each will end up with distinct naming and structure that made sense in isolation but does not fit the broader picture.

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
This adds a new memorandum GEP that attempts to chart a vision for where our TLS configuration will live and what it will be called.

**Which issue(s) this PR fixes**:
Fixes #2907

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @arkodg @candita @youngnick

### CODE REVIEW

**Comment by robscott** [comment]
Deploy Preview: https://deploy-preview-2910--kubernetes-sigs-gateway-api.netlify.app/geps/gep-2907/

File: geps/gep-2907/index.md
Comment (arkodg) [inline geps/gep-2907/index.md:85]: +1 to `Gateway.BackendTLS.ClientCertificateRef`

File: geps/gep-2907/index.md
Comment (arkodg) [inline geps/gep-2907/index.md:62]: did you mean `Listener.TLS.Frontend.Validation` ?

Comment (robscott) [reply geps/gep-2907/index.md:62]: Nope, this was intentional. Part of the idea behind this GEP is that `Listener.TLS` is really entirely focused on 1+2 - the bits of TLS between frontend and Gateway. That means that if we were to add any additional TLS configuration in the `Listerner.TLS` struct, it would tied to that limited scope, and thus we don't really need a separate `Frontend` struct.

One could make an argument that `Listener.TLS.Validation` should be the field name here to avoid any ambiguity, but in this specific context, I think that it's probably helpful to specifically spell out `frontend`.

Comment (youngnick) [reply geps/gep-2907/index.md:62]: I think that if you include basically the above comment into the text of the GEP, we should be good here.

**Comment by youngnick** [comment]
LGTM with one sadly blocking comment about explaining the naming things better.

File: geps/gep-2907/index.md
Comment (candita) [inline geps/gep-2907/index.md:89]: Could you add details on why Listener.TLS.BackendValidation wouldn't work?  Wouldn't it be analogous to Listener.TLS.FrontendValidation that was proposed?

File: geps/gep-2907/index.md
Comment (candita) [inline geps/gep-2907/index.md:16]: I see some proposed changes, so this is not really a non-goal.

**Review by candita — COMMENTED** [review]
Remembering the discussions around why the contents of BackendTLSPolicy wasn't going to be added directly to Gateway, one reason was because BackendTLS should be accessible to the app developer role, and the burden of app or service-level certificate management shouldn't be placed on the cluster operator. How do we reconcile those discussions with this proposal? Add a new role, like `application operator` or similar?

If we kept both BackendTLSPolicy and Gateway.BackendTLS, then we'd need to define how and when one could override the other. 

How about considering a new policy attachment called `FrontendTLSPolicy`, which would be available to cluster operators, but maybe not app developers. It would be similar to BackendTLSPolicy in many ways, but perhaps targetRefs could be listeners on the gateway?  Would it work for @arkodg's use cases ?

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:16]: I guess what I'm trying to say here is that if we translate this into actual API changes, those will be tracked by the GEPs associated with those APIs. This GEP is a memorandum GEP and not meant to track or own any individual API changes, will try to clean this up.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Added a note linking to the GEPs that will be affected by this one, but wanting to keep the actual ownership and detail of those changes to those GEPs.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:37]: Definitely open to better names here, but I think "paths" might unfortunately get confused with HTTP paths.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:22]: Going to come back to this one as it requires more thought. They're not necessarily non-goals, just on the periphery of what I was focused on here.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, but left "three" in for now since I don't think this number will change.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Good call, this was confusing. Reworked this a bit since I'm mostly trying to describe the concept of a Gateway when it comes to TLS to someone that might not be that familiar with the rest of the API.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Good call, added a link to our Listener Isolation docs.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:39]: Good call, reworked this to include definitions.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:102]: Added a paragraph to cover this. In general, having FrontendValidation at the Listener level allows a Gateway to accept different client certs per SNI (a matching attribute of the frontend -> Gateway connection). On the other hand, when determining the identity a Gateway should use when connecting to a backend, it should likely either be tied directly to the Gateway or Backend, but the Listener is not particularly relevant in this context.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, thanks!

**Review by arkodg — APPROVED** [review]
Looks good, thanks for adding this GEP

File: geps/gep-2907/index.md
Comment (Miciah) [reply geps/gep-2907/index.md:37]: How about "legs in a client session"?  Some alternative words: segments, links, components.

Comment (candita) [reply geps/gep-2907/index.md:37]: +1 to segments

**Comment by youngnick** [comment]
/lgtm

subject to @candita and @Miciah being satisfied with language updates. This is a 1.1 release blocker, since it blocks moving forward with GEP-91, which is also a release blocker for 1.1.

File: geps/gep-2907/index.md
Comment (dprotaso) [inline geps/gep-2907/index.md:122]: Can we settle what to do hostnames/sans?

eg. in the client validation GEP we have

```go
  // SubjectAltNames contains one or more alternate names to verify
  // the subject identity in the certificate presented by the client.
  //
  // Support: Core
  //
  // +optional
  // +kubebuilder:validation:MinItems=1
  SubjectAltNames []string `json:"subjectAltNames,omitempty"`
```

but in BackendTLSPolicy we have

```go
// Hostname is used for two purposes in the connection between Gateways and
// backends:
//
// 1. Hostname MUST be used as the SNI to connect to the backend (RFC 6066).
// 2. Hostname MUST be used for authentication and MUST match the certificate
//    served by the matching backend.
//
// Support: Core
Hostname v1beta1.PreciseHostname `json:"hostname"`
```

It'd be great to one name these things the same and potentially settle on supporting alternate names in both frontend and backend use cases.

Comment (dprotaso) [reply geps/gep-2907/index.md:122]: cc @arkodg

Comment (robscott) [reply geps/gep-2907/index.md:122]: For now I believe @arkodg has removed SANs from the client validation GEP. I think we definitely need to consider both if/when there's overlap again, I'm particularly interested in adding SANs to BackendTLSPolicy for example, but I think both are out of scope for v1.1 so we can follow up when they're proposed in the future.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Good call, added, thanks! I also reordered the sections here so they're hopefully a bit less confusing.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:22]: Added an "Out of initial scope" section below this to cover this topic for now.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:86]: Copied nearly verbatim, thanks!

**Comment by robscott** [comment]
Thanks everyone for the feedback! I think I've addressed everything now, so PTAL. I'm going to remove the hold so the next LGTM merges this. I think we've got pretty good consensus around this GEP at this point and I really want to unblock @arkodg's GEP that depends on this one.

/hold cancel

File: geps/gep-2907/index.md
Comment (candita) [reply geps/gep-2907/index.md:22]: I called it "Longer Term Goals" in GEP-1897, but they mean the same thing.

**Comment by candita** [comment]
Sorry, I feel like we are rushing this.

/hold

File: geps/gep-2907/index.md
Comment (frankbu) [inline geps/gep-2907/index.md:56]: For `similar to Envoy` terminology, 3+4 are the upstream, 1+2 is downstream. I see what you mean by "avoid confusion" 😃 
```suggestion
A guiding principle in this naming is to use consistent naming for “Upstream”
(3+4) and “Downstream” (1+2), similar to Envoy. To avoid the confusion with what
```

File: geps/gep-2907/index.md
Comment (candita) [inline geps/gep-2907/index.md:127]: @robscott I feel we shouldn't add certain backend configuration that the app developer should have access to.  This part should be left TBD if it's not needed to unblock GEP-91.  We should not have a proposal that includes the indecision of "either or both" right from the beginning.

I mentioned earlier:

> Remembering the discussions around why the content of BackendTLSPolicy wasn't going to be added directly to Gateway, one reason was because BackendTLS should be accessible to the app developer role, and the burden of app or service-level certificate management shouldn't be placed on the cluster operator. How do we reconcile those discussions with this proposal? Add a new role, like application operator or similar?


>How about considering a new policy attachment called FrontendTLSPolicy, which would be available to cluster operators, but maybe not app developers. It would be similar to BackendTLSPolicy in many ways, but perhaps targetRefs could be listeners on the gateway? Would it work for @arkodg's use cases ?

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: 🤦  updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:127]: Good question, I think there are several reasons BackendTLSPolicy exists as a separate policy resource:

1. Some parts of the config, such as the SNI that should be used to connect to the Service/backend only make sense when attached per Service - it's very unlikely that you'd want to use the same SNI to connect to all backends from a Gateway.
2. The Service API is overloaded and very slow to change, Gateway API has taken the approach of exposing config at that level via policy resources that are tied to specific personas (why Session Persistence config and Backend TLS are in different resources).

I don't think either of those reasons really apply to frontend TLS. Instead, frontend TLS could theoretically be configured either for an entire Gateway, or a specific Gateway listener. Given that many implementations already support distinguishing this config per SNI, it seems to make sense to start with listener level attachment. I think that the persona that would be responsible for a Gateway is not sufficiently different than the persona that would be responsible for frontend TLS, so I tend to think that the current proposal is the best option available to us.

I'll work on clarifying all of that in the GEP.

File: geps/gep-2907/index.md
Comment (candita) [inline geps/gep-2907/index.md:57]:
```suggestion
is upstream and downstream, Gateway API will use “Frontend” (1+2) and “Backend” (3+4).
```

File: geps/gep-2907/index.md
Comment (candita) [inline geps/gep-2907/index.md:80]: I'd like to summarize the proposed segments, only because when I look at them together they don't seem as unified as they could be:

> Gateway.Listener.TLS.FrontendValidation
> Gateway.Listener.TLS
> Gateway.Spec.BackendTLS.ClientCertificateRef
> Gateway.Spec.BackendTLS.Validation

To me the problem of dealing with Listener is better solved by doing something like this:

> Gateway.Listener.TLS.FrontendValidation
> Gateway.Listener.TLS
> Gateway.Transmitter.TLS.BackendValidation
> Gateway.Transmitter.TLS
 
or even

> Gateway.Listener.TLS.Validation
> Gateway.Listener.TLS
> Gateway.Transmitter.TLS.Validation
> Gateway.Transmitter.TLS

We don't have to use the term `Transmitter`, just something other than `BackendTLS`, which clashes with `BackendTLSPolicy`.  I feel a move toward this pattern presents an easier learning curve for people with memory challenges (like me).

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:151]: Added a new section "Why Not FrontendTLSPolicy?" to address this.

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md]: Good call, updated, thanks!

File: geps/gep-2907/index.md
Comment (robscott) [reply geps/gep-2907/index.md:81]: Good questions, I've added two separate but related sections to address this:

* Why the inconsistency between Frontend and Backend TLS config on Gateways?
* Why Not Listener level to match FrontendValidation?

I've also added a note that we should reconsider these answers before developing any kind of backend TLS config at a Gateway level. Hope that all works for you.

**Comment by robscott** [comment]
Thanks for all the great feedback @candita, I think I've addressed everything now. Would appreciate an additional round of review and/or LGTM from anyone that has time so we can get this one in and unblock the dependent updates.

/hold cancel

**Review by arkodg — APPROVED** [review]
/lgtm

**Pull request merged**
